File: 268.missing-number.cpp
/*
 * @lc app=leetcode id=268 lang=cpp
 *
 * [268] Missing Number
 */

#include <bits/stdc++.h>

#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  int missingNumber(vector<int>& nums) {
    int len = nums.size();
    sort(nums.begin(), nums.end());
    for (int i = 0; i < len; i++) {
      if (i != nums[i]) {
        return i;
      }
    }
    return len;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> nums = {3, 0, 1};
  assert(s.missingNumber(nums) == 2);
  nums = {0, 1};
  assert(s.missingNumber(nums) == 2);
  nums = {9, 6, 4, 2, 3, 5, 7, 0, 1};
  assert(s.missingNumber(nums) == 8);
  nums = {0};
  assert(s.missingNumber(nums) == 1);
  return 0;
}

File: 268.missing-number.rs
/*
 * @lc app=leetcode id=268 lang=rust
 *
 * [268] Missing Number
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn missing_number(nums: Vec<i32>) -> i32 {
        let mut count = vec![false; nums.len() + 1];
        for n in nums {
            count[n as usize] = true;
        }
        for (idx, has) in count.iter().enumerate() {
            if !has {
                return idx as i32;
            }
        }
        -1
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::missing_number(vec![3, 0, 1]), 2);
}

#[test]
fn test2() {
    assert_eq!(Solution::missing_number(vec![0, 1]), 2);
}

#[test]
fn test3() {
    assert_eq!(Solution::missing_number(vec![9, 6, 4, 2, 3, 5, 7, 0, 1]), 8);
}
